packages/zoe/src/contracts/loan/liquidate.js

```diff
@@ -3,81 +3,56 @@ import '../../../exported';
 
 import { E } from '@agoric/eventual-send';
 
-import { depositToSeat, withdrawFromSeat } from '../../contractSupport';
+import { offerTo } from '../../contractSupport/zoeHelpers';
 
 export const doLiquidation = async (
   zcf,
   collateralSeat,
   autoswapPublicFacetP,
   lenderSeat,
 ) => {
-  const zoeService = zcf.getZoeService();
   const loanMath = zcf.getTerms().maths.Loan;
 
   const allCollateral = collateralSeat.getAmountAllocated('Collateral');
 
-  const { Collateral: collateralPayment } = await withdrawFromSeat(
-    zcf,
-    collateralSeat,
-    {
-      Collateral: allCollateral,
-    },
-  );
+  const swapInvitation = E(autoswapPublicFacetP).makeSwapInInvitation();
+
+  const toAmounts = harden({ In: allCollateral });
 
   const proposal = harden({
-    give: { In: allCollateral },
+    give: toAmounts,
     want: { Out: loanMath.getEmpty() },
   });
 
-  const payments = harden({ In: collateralPayment });
-
-  const swapInvitation = E(autoswapPublicFacetP).makeSwapInInvitation();
+  const keywordMapping = harden({
+    Collateral: 'In',
+    Loan: 'Out',
+  });
 
-  const autoswapUserSeat = E(zoeService).offer(
+  const { userSeatPromise: autoswapUserSeat, deposited } = await offerTo(
+    zcf,
     swapInvitation,
+    keywordMapping,
     proposal,
-    payments,
+    collateralSeat,
+    lenderSeat,
   );
 
-  /**
-   * @param {{ Out: Promise<Payment>; In: Promise<Payment>; }} payouts
-   */
-  const handlePayoutsAndShutdown = async payouts => {
-    const { Out: loanPayout, In: collateralPayout } = payouts;
-    const { Out: loanAmount, In: collateralAmount } = await E(
-      autoswapUserSeat,
-    ).getCurrentAllocation();
-    // AWAIT ///
-    const amounts = harden({
-      Collateral: collateralAmount,
-      Loan: loanAmount,
-    });
-    const payoutPayments = harden({
-      Collateral: collateralPayout,
-      Loan: loanPayout,
-    });
-    await depositToSeat(zcf, lenderSeat, amounts, payoutPayments);
-
-    const closeSuccessfully = () => {
-      lenderSeat.exit();
-      collateralSeat.exit();
-      zcf.shutdown('your loan had to be liquidated');
-    };
-
-    const closeWithFailure = err => {
-      lenderSeat.kickOut(err);
-      collateralSeat.kickOut(err);
-      zcf.shutdownWithFailure(err);
-    };
-
-    await E(autoswapUserSeat)
-      .getOfferResult()
-      .then(closeSuccessfully, closeWithFailure);
+  const closeSuccessfully = () => {
+    lenderSeat.exit();
+    collateralSeat.exit();
+    zcf.shutdown('your loan had to be liquidated');
+  };
+
+  const closeWithFailure = err => {
+    lenderSeat.fail(err);
+    collateralSeat.fail(err);
+    zcf.shutdownWithFailure(err);
   };
 
-  return E(autoswapUserSeat)
-    .getPayouts()
-    .then(handlePayoutsAndShutdown);
+  const offerResultP = E(autoswapUserSeat).getOfferResult();
+  await deposited;
+  offerResultP.then(closeSuccessfully, closeWithFailure);
```

Comment:
Any reason not to use `Promise.all([deposited, offerResult]).then(`?

Reply:
We have to ensure the deposit on the lenderSeat is done before the lenderSeat is exited.

Reply:
My point was that in both cases, we wait for both promises to complete. The present code waits for them in order, but doesn't force the `deposit` to happen before the offer. The offerResult promise was created earlier, and its execution is unblocked.

The reason I suggested the change is because the consistent treatment of the two promises that we're waiting for with a single `Promise.all()` seems cleaner syntactically.

The only difference in execution that I can see is that if `deposited` fails, the present code wouldn't call `closeWithFailure`, while the combined version would. I presume this would actually be an improvement, and the reason it wasn't done initially was that a failure in the deposit was seen as unlikely.

Reply:
No, it's not about waiting for the `offerResult`, it's that the present case forces the `deposit` to happen before the `lenderSeat` is exited. It's important that we keep that.

Reply:
Hmm, let me think about this.

Reply:
Ah, ok, I think I figured out what is going on here. You're right that in the case that both promises resolve, we wait for `deposited` to resolve before the `lenderSeat` exits, so that case works. However, when the `offerHandler` throws, we try to exit the `lenderSeat` before it is deposited to. This is because Promise.all() rejects immediately if any of the promises reject individually, so what happens if we have `Promise.all([deposited, offerResult]).then(` is that we do not await `deposited` before trying to exit the lenderSeat in `closeWithFailure`. But, we do want to always `deposit` first, so that's not the right behavior. 

We want to `await deposited` regardless of whether `offerResultP` rejects or not, so I think the original version is the better option.

Reply:
What should happen if `deposited` rejects?

Reply:
What should happen if `deposited` never settles?

Reply:
Good questions. If `deposited` rejects, then we have profoundly misunderstood something and it's unclear how we should recover. This would mean that we could not deposit the payouts (refund or otherwise) to the lenderSeat. We do have access to the payouts through the `autoswapUserSeat` but it's not clear how we would deposit again in a way that would be successful given that our earlier attempt failed. The reason why `offerResultP` rejecting is handled better is that it assumes that `deposited` likely resolves due to a refund payout, and we can safely exit `lenderSeat` because it got the refund. 

Likewise, if `deposited` never settles, something has gone horribly wrong, because 1) `autoswap` always exits the seat quickly and 2) even in the case of error in autoswap, we should still get a payout that we can deposit safely. Without having deposited a refund successfully it is unclear how we should recover. For instance, we don't want to exit seats or shutdown the contract if only the contract has access to payments.

Reply:
Let me know if any of these conversations should hold up the PR. Chris has already approved and I think it's ready to merge once the CI passes

Reply:
*If* our definition of issuer misbehavior includes rejection or never settling under these conditions as, then the behavior you describe is consistent with our threat model for payout liveness. For any given seat, a misbehaving issuer *among the issuers that this seat explicitly mentions* in its offer, a misbehaving issuer can block payout liveness for the seat as a whole. This is just another instance of that same allowed failure mode. The security claim we need to make for payout liveness in the face of issuer misbehavior is only

A misbehaving issuer cannot block payout liveness for seats that do not mention that issuer *as an issuer* in their offer.

Btw, we need to gather such precise statements of the security claims we are making, both for what they include and what they exclude.

Reply:
> Let me know if any of these conversations should hold up the PR. Chris has already approved and I think it's ready to merge once the CI passes

I still see no reason for this to block on me. Please proceed.

Reply:
> Btw, we need to gather such precise statements of the security claims we are making, both for what they include and what they exclude.

For this claim, already done though it should be refined.

https://github.com/Agoric/agoric-sdk/wiki/Limited-Risk-from-Bad-Issuers